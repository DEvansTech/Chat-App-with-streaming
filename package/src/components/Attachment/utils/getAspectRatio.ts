import type { Attachment } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../../types/types';

/**
 * Returns the aspect ratio of an image attachment.
 *
 * @param image Image attachment.
 * @returns {number}
 */
export function getAspectRatio<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(image: Attachment<StreamChatGenerics>) {
  if (image.type !== 'image') {
    throw new Error('getAspectRatio() can only be called on an image attachment');
  }

  if (!image.original_width || !image.original_height) return 1;

  return image.original_width / image.original_height;
}
