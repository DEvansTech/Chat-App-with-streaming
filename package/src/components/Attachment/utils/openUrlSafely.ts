import { Linking } from 'react-native';

export const openUrlSafely = (url?: string) => {
  let finalUrl = url as string;
  const pattern = new RegExp(/^\S+:\/\//);

  if (!pattern.test(finalUrl)) {
    finalUrl = 'http://' + url;
  }

  Linking.canOpenURL(finalUrl).then((supported) => {
    if (supported) {
      Linking.openURL(finalUrl);
    } else {
      console.warn(`Don't know how to open URI: ${finalUrl}`);
    }
  });
};
