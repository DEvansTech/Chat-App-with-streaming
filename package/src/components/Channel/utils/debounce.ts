/* eslint-disable @typescript-eslint/no-explicit-any */
import type { DebounceSettings } from 'lodash';
import debounce from 'lodash/debounce';

export const heavyDebounce = <T extends (...args: any) => any>(
  callback: T,
  ms = 500,
  option: DebounceSettings = {
    leading: true,
    trailing: true,
  },
) => debounce(callback, ms, option);
