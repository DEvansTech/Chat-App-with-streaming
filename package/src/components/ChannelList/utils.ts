import type { Channel, StreamChat } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../types/types';

type MoveParameters<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = {
  channels: Channel<StreamChatClient>[];
  cid: string;
};

export const moveChannelUp = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  cid,
  channels = [],
}: MoveParameters<StreamChatClient>) => {
  // get channel from channels
  const index = channels.findIndex((c) => c.cid === cid);
  if (index <= 0) return channels;
  const channel = channels[index];

  // remove channel from current position and add to start
  channels.splice(index, 1);
  channels.unshift(channel);

  return [...channels];
};

type GetParameters<StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics> =
  {
    client: StreamChat<StreamChatClient>;
    id: string;
    type: string;
  };

export const getChannel = async <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  client,
  id,
  type,
}: GetParameters<StreamChatClient>) => {
  const channel = client.channel(type, id);
  await channel.watch();
  return channel;
};

export const DEFAULT_QUERY_CHANNELS_LIMIT = 10;
export const MAX_QUERY_CHANNELS_LIMIT = 30;
