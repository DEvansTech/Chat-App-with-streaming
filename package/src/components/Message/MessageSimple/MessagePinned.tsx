import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { Pin } from '../../../icons/Pin';

import {
  MessageContextValue,
  useMessageContext,
} from '../../../contexts/messageContext/MessageContext';
import { useTheme } from '../../../contexts/themeContext/ThemeContext';
import { useTranslationContext } from '../../../contexts/translationContext/TranslationContext';

const styles = StyleSheet.create({
  text: {
    marginLeft: 5,
  },
  view: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 8,
    marginLeft: 30,
    marginTop: 5,
  },
});

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import { useChatContext } from '../../../contexts/chatContext/ChatContext';

export type MessagePinnedPropsWithContext<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
> = Pick<MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'alignment' | 'message'>;

const MessagePinnedWithContext = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
>(
  props: MessagePinnedPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { alignment, message } = props;
  const {
    theme: {
      colors: { grey },
    },
  } = useTheme();
  const { t } = useTranslationContext();
  const { client } = useChatContext();
  return (
    <View
      style={[
        styles.view,
        {
          justifyContent: alignment === 'left' ? 'flex-start' : 'flex-end',
        },
      ]}
      testID='message-pinned'
    >
      <Pin height={16} pathFill={grey} width={24} />
      <Text style={[{ color: grey }, styles.text]}>
        Pinned by{' '}
        {message?.pinned_by?.id === client?.user?.id ? t('You') : message?.pinned_by?.name}
      </Text>
    </View>
  );
};

const areEqual = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  prevProps: MessagePinnedPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
  nextProps: MessagePinnedPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { message: prevMessage } = prevProps;
  const { message: nextMessage } = nextProps;
  const messageEqual =
    prevMessage.deleted_at === nextMessage.deleted_at &&
    prevMessage.status === nextMessage.status &&
    prevMessage.type === nextMessage.type &&
    prevMessage.text === nextMessage.text &&
    prevMessage.pinned === nextMessage.pinned;
  if (!messageEqual) return false;
  return true;
};

const MemoizedMessagePinned = React.memo(
  MessagePinnedWithContext,
  areEqual,
) as typeof MessagePinnedWithContext;

export type MessagePinnedProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
> = Partial<MessagePinnedPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>>;

export const MessagePinned = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
>(
  props: MessagePinnedProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { alignment, lastGroupMessage, message } = useMessageContext<At, Ch, Co, Ev, Me, Re, Us>();

  return (
    <MemoizedMessagePinned
      {...{
        alignment,
        lastGroupMessage,
        message,
      }}
      {...props}
    />
  );
};

MessagePinned.displayName = 'MessagePinned{messageSimple{pinned}}';
