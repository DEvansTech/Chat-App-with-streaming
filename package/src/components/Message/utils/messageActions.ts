import type { ExtendableGenerics } from 'stream-chat';

import type { MessageContextValue } from '../../../contexts/messageContext/MessageContext';
import type { OwnCapabilitiesContextValue } from '../../../contexts/ownCapabilitiesContext/OwnCapabilitiesContext';
import type { DefaultStreamChatGenerics } from '../../../types/types';
import type { MessageActionType } from '../../MessageOverlay/MessageActionListItem';

export type MessageActionsParams<
  StreamChatClient extends ExtendableGenerics = DefaultStreamChatGenerics,
> = {
  blockUser: MessageActionType;
  copyMessage: MessageActionType;
  deleteMessage: MessageActionType;
  dismissOverlay: () => void;
  editMessage: MessageActionType;
  error: boolean | Error;
  flagMessage: MessageActionType;
  isThreadMessage: boolean;
  messageReactions: boolean;
  muteUser: MessageActionType;
  ownCapabilities: OwnCapabilitiesContextValue;
  pinMessage: MessageActionType;
  quotedReply: MessageActionType;
  retry: MessageActionType;
  threadReply: MessageActionType;
  unpinMessage: MessageActionType;
} & Pick<MessageContextValue<StreamChatClient>, 'message' | 'isMyMessage'>;

export type MessageActionsProp<
  StreamChatClient extends ExtendableGenerics = DefaultStreamChatGenerics,
> = (param: MessageActionsParams<StreamChatClient>) => MessageActionType[];

export const messageActions = <
  StreamChatClient extends ExtendableGenerics = DefaultStreamChatGenerics,
>({
  blockUser,
  copyMessage,
  deleteMessage,
  editMessage,
  error,
  flagMessage,
  isMyMessage,
  isThreadMessage,
  message,
  messageReactions,
  ownCapabilities,
  pinMessage,
  quotedReply,
  retry,
  threadReply,
  unpinMessage,
}: MessageActionsParams<StreamChatClient>) => {
  if (messageReactions) {
    return undefined;
  }

  const actions: Array<MessageActionType | null> = [];

  if (error && isMyMessage) {
    actions.push(retry);
  }

  if (ownCapabilities.quoteMessage && !isThreadMessage && !error) {
    actions.push(quotedReply);
  }

  if (ownCapabilities.sendReply && !isThreadMessage && !error) {
    actions.push(threadReply);
  }

  if (
    (isMyMessage && ownCapabilities.updateOwnMessage) ||
    (!isMyMessage && ownCapabilities.updateAnyMessage)
  ) {
    actions.push(editMessage);
  }

  if (message.text && !error) {
    actions.push(copyMessage);
  }

  if (!isMyMessage && ownCapabilities.flagMessage) {
    actions.push(flagMessage);
  }

  if (ownCapabilities.pinMessage && !message.pinned) {
    actions.push(pinMessage);
  }

  if (ownCapabilities.pinMessage && message.pinned) {
    actions.push(unpinMessage);
  }

  if (!isMyMessage && ownCapabilities.banChannelMembers) {
    actions.push(blockUser);
  }

  if (
    (isMyMessage && ownCapabilities.deleteOwnMessage) ||
    (!isMyMessage && ownCapabilities.deleteAnyMessage)
  ) {
    actions.push(deleteMessage);
  }

  return actions;
};
