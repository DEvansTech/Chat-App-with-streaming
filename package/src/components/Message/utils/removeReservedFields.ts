import type { DefaultStreamChatGenerics } from '../../../types/types';
import type { MessageType } from '../../MessageList/hooks/useMessageList';

export const removeReservedFields = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  message: MessageType<StreamChatClient>,
) => {
  const retryMessage = { ...message };
  const reserved = [
    'cid',
    'config',
    'created_at',
    'created_by',
    'id',
    'last_message_at',
    'member_count',
    'type',
    'updated_at',
  ];
  reserved.forEach((key) => {
    delete retryMessage[key];
  });

  return retryMessage;
};
