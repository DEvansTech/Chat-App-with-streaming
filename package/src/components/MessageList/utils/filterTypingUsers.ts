import type { ChatContextValue } from '../../../contexts/chatContext/ChatContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type { TypingContextValue } from '../../../contexts/typingContext/TypingContext';
import type { DefaultStreamChatGenerics } from '../../../types/types';

type FilterTypingUsersParams<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Pick<TypingContextValue<StreamChatClient>, 'typing'> &
  Pick<ChatContextValue<StreamChatClient>, 'client'> &
  Pick<ThreadContextValue<StreamChatClient>, 'thread'>;

export const filterTypingUsers = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  client,
  thread,
  typing,
}: FilterTypingUsersParams<StreamChatClient>) => {
  const nonSelfUsers: string[] = [];

  if (!client || !client.user || !typing) return nonSelfUsers;

  const typingKeys = Object.keys(typing);

  typingKeys.forEach((typingKey) => {
    if (!typing[typingKey]) return;

    /** removes own typing events */
    if (client.user?.id === typing[typingKey].user?.id) {
      return;
    }

    const isRegularEvent = !typing[typingKey].parent_id && !thread?.id;
    const isCurrentThreadEvent = typing[typingKey].parent_id === thread?.id;

    /** filters different threads events */
    if (!isRegularEvent && !isCurrentThreadEvent) {
      return;
    }

    const user = typing[typingKey].user?.name || typing[typingKey].user?.id;
    if (user) {
      nonSelfUsers.push(user);
    }
  });

  return nonSelfUsers;
};
