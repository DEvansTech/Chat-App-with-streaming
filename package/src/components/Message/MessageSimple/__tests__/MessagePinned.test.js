import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import { MessagePinned } from '../MessagePinned';

import { ThemeProvider } from '../../../../contexts/themeContext/ThemeContext';
import { defaultTheme } from '../../../../contexts/themeContext/utils/theme';
import {
  generateMessage,
  generateStaticMessage,
} from '../../../../mock-builders/generator/message';
import { generateStaticUser } from '../../../../mock-builders/generator/user';

afterEach(cleanup);

describe('MessagePinned', () => {
  it('should render message pinned', async () => {
    const staticUser = generateStaticUser(0);
    const message = generateMessage({
      user: { ...staticUser, image: undefined },
    });
    const { getByTestId, rerender, toJSON } = render(
      <ThemeProvider style={defaultTheme}>
        <MessagePinned alignment='right' message={message} />
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(getByTestId('message-pinned')).toBeTruthy();
    });

    rerender(
      <ThemeProvider style={defaultTheme}>
        <MessagePinned alignment='right' message={message} />
      </ThemeProvider>,
    );

    const staticMessage = generateStaticMessage('hi', {
      user: staticUser,
    });

    rerender(
      <ThemeProvider style={defaultTheme}>
        <MessagePinned alignment='left' message={staticMessage} />
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(getByTestId('message-pinned')).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
