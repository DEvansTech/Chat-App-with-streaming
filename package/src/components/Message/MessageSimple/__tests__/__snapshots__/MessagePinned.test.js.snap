// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagePinned should render message pinned 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flexDirection": "row",
        "marginBottom": 8,
        "marginLeft": 30,
        "marginTop": 5,
      },
      Object {
        "justifyContent": "flex-start",
      },
    ]
  }
  testID="message-pinned"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={16}
    bbWidth={24}
    focusable={false}
    height={16}
    meetOrSlice={0}
    minX={0}
    minY={0}
    pathFill="#7A7A7A"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 16,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={16}
    width={24}
  >
    <RNSVGGroup>
      <RNSVGPath
        clipRule={0}
        d="M16.041 0.293372C15.7956 0.0479219 15.4405 -0.0529382 15.1027 0.0268118C14.7649 0.106562 14.4924 0.355552 14.3826 0.684852L13.7492 2.58516L9.0804 7.25401C6.69213 6.71299 3.39999 8.68517 1.29335 10.7918C0.902218 11.1829 0.902218 11.8171 1.29335 12.2082L5.83434 16.7492L1.70973 20.8738C1.3186 21.265 1.3186 21.8991 1.70973 22.2902C2.10086 22.6814 2.73501 22.6814 3.12614 22.2902L7.25074 18.1656L11.7918 22.7067C12.1829 23.0978 12.8171 23.0978 13.2082 22.7067C15.3148 20.6 17.287 17.3079 16.746 14.9196L21.4148 10.2508L23.3151 9.61735C23.6444 9.50758 23.8934 9.23514 23.9732 8.89731C24.0529 8.55948 23.9521 8.20444 23.7066 7.95899L16.041 0.293372ZM15.5748 3.4429L15.7705 2.85567L21.1443 8.2295L20.5571 8.42525C20.4096 8.47441 20.2756 8.55725 20.1656 8.6672L14.9164 13.9164C14.6481 14.1847 14.5545 14.5814 14.6744 14.9413C15.1518 16.3733 13.8573 18.8847 12.4693 20.5514L3.44861 11.5307C5.11529 10.1427 7.62666 8.84822 9.0587 9.3256C9.4186 9.4455 9.8153 9.3519 10.0836 9.0836L15.3328 3.83438C15.4427 3.72444 15.5256 3.59041 15.5748 3.4429Z"
        fill={4286216826}
        fillRule={0}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      Array [
        Object {
          "color": "#7A7A7A",
        },
        Object {
          "marginLeft": 5,
        },
      ]
    }
  >
    Pinned by
     
    You
  </Text>
</View>
`;
